nn/conformer.py

```diff
@@ -0,0 +1,208 @@
+"""
+Conformer code.
+Ref: https://arxiv.org/abs/2005.08100
+"""
+
+from typing import Tuple, List, Union
+from .. import nn
+
+
+class _PositionwiseFeedForward(nn.Module):
+  """
+  Conformer position-wise feedforward neural network layer
+      FF -> Activation -> Dropout -> FF
+  """
+
+  def __init__(self, out_dim: int, dim_ff: int, dropout: float, activation):
+    """
+    :param out_dim:
+    :param dim_ff:
+    :param dropout:
+    :param activation:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+    self.activation = activation
+
+    self.linear_ff = nn.Linear(n_out=dim_ff)
+    self.linear_out = nn.Linear(n_out=out_dim)
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    x_ff1 = self.linear_ff(inp)
+    x_act = self.activation(x_ff1)
+    x_drop = nn.dropout(x_act, dropout=self.dropout)
+    x_ff2 = self.linear_out(x_drop)
+    return x_ff2
+
+
+class _ConformerConvBlock(nn.Module):
+  """
+  Conformer convolution block
+      FF -> GLU -> depthwise conv -> BN -> Swish -> FF
+  """
+
+  def __init__(self, out_dim: int, kernel_size: int, batch_norm_eps: float = 1e-5,
+      batch_norm_momentum: float = 0.1, batch_norm_other_opts=None):
+    """
+    :param out_dim:
+    :param kernel_size:
+    """
+    super().__init__()
+
+    self.positionwise_conv1 = nn.Linear(n_out=out_dim * 2)
+    self.depthwise_conv = nn.Conv(n_out=out_dim, filter_size=(kernel_size,), groups=out_dim, padding='same')
+    self.positionwise_conv2 = nn.Linear(n_out=out_dim)
+
+    if batch_norm_other_opts is None:
+      batch_norm_other_opts = {}
+    self.batch_norm = nn.BatchNorm(
+      epsilon=batch_norm_eps, momentum=batch_norm_momentum, update_sample_only_in_training=True,
+      delay_sample_update=True, **batch_norm_other_opts)
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    x_conv1 = self.positionwise_conv1(inp)
+    x_act = nn.glu(x_conv1)
+    x_depthwise_conv = self.depthwise_conv(x_act)
+    x_bn = self.batch_norm(x_depthwise_conv)
+    x_swish = nn.swish(x_bn)
+    x_conv2 = self.positionwise_conv2(x_swish)
+    return x_conv2
+
+
+class _ConformerConvSubsampleLayer(nn.Module):
+  """
+  Conv 2D block with optional max-pooling
+  """
+
+  def __init__(self, filter_sizes: List[Tuple[int, ...]], pool_sizes: Union[List[Tuple[int, ...]], None],
```

Comment:
`None` is not the correct type. But anyway use `Optional[...]` instead.

Reply:
You don't actually expect `Tuple[int, ...]` for the filter sizes and pool sizes, right? You always expect 1D convolution and pooling, so it would be `Tuple[int]` (a single int). But also, with #61 and when you use `nn.Conv1d`, you can simply use a single `int` instead.

Reply:
This is 2D Conv so a tuple of ints is expected.

Reply:
But then specify it explicitly: `Tuple[int,int]`

And really? This is 2D conv?

Reply:
yes. Also similar to ESPNet:
- https://github.com/espnet/espnet/blob/master/espnet2/asr/encoder/conformer_encoder.py#L145
- https://github.com/espnet/espnet/blob/a1a04e05a828353c74775327095f78273faba4f3/espnet/nets/pytorch_backend/transformer/subsampling.py#L59